File: Minimal/Clojure/url_shortener.clj
(ns url-shortener.core)

(defn generate-short-code [length]
  (let [characters "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"]
    (->> characters
         (repeatedly #(rand-nth %))
         (take length)
         (apply str))))

(defn create-shortener []
  (atom {}))

(defn shorten! [shortener url]
  (let [short (generate-short-code 6)]
    (swap! shortener assoc short url)
    short))

(defn get-url [shortener short]
  (get @shortener short))

;; Example usage
(defn -main []
  (let [shortener (create-shortener)
        short (shorten! shortener "https://www.example.com")]
    (println (str "Short URL: " short))
    (if-let [original (get-url shortener short)]
      (println (str "Original URL: " original))
      (println "URL not found"))))

(-main)

File: Minimal/R/url_shortener.R
# URL Shortener in R

URLShortener <- function() {
  urls <- list()
  
  list(
    shorten = function(url) {
      # Generate random 6-character short code
      short <- paste0(sample(c(letters, LETTERS, 0:9), 6, replace = TRUE), collapse = "")
      urls[[short]] <<- url
      return(short)
    },
    
    get_url = function(short) {
      if (short %in% names(urls)) {
        return(urls[[short]])
      } else {
        return(NULL)
      }
    }
  )
}

# Example usage
shortener <- URLShortener()
short <- shortener$shorten("https://www.example.com")
cat("Short URL:", short, "\n")

original <- shortener$get_url(short)
if (!is.null(original)) {
  cat("Original URL:", original, "\n")
} else {
  cat("URL not found\n")
}
